proto/tendermint/abci/types.proto

```diff
@@ -117,27 +122,60 @@ message RequestApplySnapshotChunk {
   string sender = 3;
 }
 
+message RequestPrepareProposal {
+  bytes                   hash                 = 1;
+  tendermint.types.Header header               = 2 [(gogoproto.nullable) = false];
+  // txs is an array of transactions that will be included in a block,
+  // sent to the app for possible modifications.
+  repeated bytes          txs                  = 3;
+  LastCommitInfo          last_commit_info     = 4 [(gogoproto.nullable) = false];
+  repeated Evidence       byzantine_validators = 5 [(gogoproto.nullable) = false];
+  // the modified transactions cannot exceed this size.
+  int64                   max_tx_bytes         = 6;
+}
+
+message RequestProcessProposal {
+  bytes                   hash                 = 1;
+  tendermint.types.Header header               = 2 [(gogoproto.nullable) = false];
```

Comment:
(further thoughts, non-blocking)

Doesn't `tendermint.types` also depend on `tendermint.abci`? I realize the code generator hides the cyclic dependency from the build tool, but I feel like we should try to make the dependencies go in only one direction if we can.

(The one case I know about specifically is `Event`, and maybe the easy solution there would be to move that over to `tendermint.types` and reverse the dependency)

Reply:
That's an excellent point. I hadn't realized about the dependency cycle. Even if the build tool is able to cope with it, I agree it's not (conceptually) clean.

The way I see it, `tendermint.abci` should depend on `tendermint.types`, and not the contrary, as `tendermint.types` seems to be defining the most basic types. But I might have got it wrong...

I'll open an issue on this, but won't include it in the ABCI++ project. I did a quick search on the file and there are preexisting references to (e.g. RequestInitChain, RequestBeginBlock)

Reply:
> The way I see it, `tendermint.abci` should depend on `tendermint.types`, and not the contrary, as `tendermint.types` seems to be defining the most basic types.

I tend to agree. That said: I think it's mainly a matter of understandability rather than correctness.

If we don't want to do it as part of ABCI++ (which is a very reasonable choice), it might not be worth bothering. Once the message type tags get buried in databases and APIs, it'll be hard to change them without breaking stuff, and I'd probably not ship a breaking change just to fix it.